### PR TITLE
Apply proposed changes by @RDouglasSharp from issue #57 #58

### LINKED ISSUE #58: Cached KVs not implemented on Adapter causing errors.
Adapter inherits most methods from the BaseModel.  The Adapter's init method, Block's forward method and CausalSelfAttention's forward method didn't implement the Cached KVs logics, causing errors.

The errors are mostly from forward method in Adapter.
Example:
```python
AttributeError: 'Parrot' object has no attribute 'rope_cache' 
TypeError: Block.forward() takes 2 positional arguments but 7 were given
```

We should add 

- rope_cache, mask_cache, kv_caches attributes to Adapter init method.
- rope, mask, max_seq_length, input_pos, kv_cache to input of Block and CausalSelfAttention forward method
- return kv_cache in  Block and CausalSelfAttention forward method

### PR DESCRIPTION
**Pull request by ArturK-85**
I apply proposed changes and tested it work good!

Fixes #58 #59 #61

### CODE REVIEW

File: lit_parrot/adapter.py [hunk 1]
```diff
@@ -19,6 +19,10 @@
 from lit_parrot.model import MLP, Parrot as BaseModel, build_rope_cache, apply_rope
 
 
+RoPECache = Tuple[torch.Tensor, torch.Tensor]
+KVCache = Tuple[torch.Tensor, torch.Tensor]
```
Comment (carmocca) [inline lit_parrot/adapter.py:23]: You can import this from the base model

File: lit_parrot/config.py [hunk 2]
```diff
@@ -24,7 +24,9 @@ def __post_init__(self):
 
     @classmethod
     def from_name(cls, name: str, **kwargs: Any) -> Self:
-        return cls(**configs[name], **kwargs)
+        conf_dict = configs[name].copy()
+        conf_dict.update(kwargs)
+        return cls(**conf_dict)
```
Comment (carmocca) [inline lit_parrot/config.py:29]: Can you add a test for this fix?

**Review by carmocca — APPROVED** [review]
LGTM

File: lit_parrot/config.py (hunk 2 above)
Comment (lantiga) [reply lit_parrot/config.py:29]: Done

File: lit_parrot/adapter.py [hunk 3]
```diff
@@ -62,39 +74,37 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
-            prefix = self.adapter_wte.weight.reshape(1, self.adapter_prompt_length, self.n_embd)
-
-            aT = prefix.size(1)
-            _, ak, av = self.attn(prefix).split(self.n_embd, dim=2)  # mayby dim=2
-            ak = ak.view(1, aT, self.n_head, head_size).repeat(B, 1, 1, 1).transpose(1, 2)
-            av = av.view(1, aT, self.n_head, head_size).repeat(B, 1, 1, 1).transpose(1, 2)
+        if kv_cache is not None:
+            cache_k, cache_v = kv_cache
+            cache_k = cache_k.type(torch.bfloat16)
+            cache_v = cache_v.type(torch.bfloat16)
```
[the remark this answers is not in the document]
Comment (lantiga) [reply lit_parrot/adapter.py]: Yep, this was indeed removed in the current PR

**Comment by ArturK-85** [comment]
Thank you @lantiga !!!

**Pull request merged**